File: aws/api_tests.py
"""
api_tests.py

AWS API Gateway Test Suite for Browses.

@dependencies:
    - requests
@run:
    - python api_tests.py
"""
import requests
import unittest

BROWSE_URL = "https://f7mlijh134.execute-api.eu-west-1.amazonaws.com/beta"
LOGINN_URL = "https://7ibd5w7y69.execute-api.eu-west-1.amazonaws.com/beta"


class TestBrowsesAuth(unittest.TestCase):
    """
    Test authenticated resources, using token from loginns.
    Check API's basic functionality and response content.
    """
    @classmethod
    def setUpClass(cls):
        cls.user = "joebloggs"
        cls.pwd = "salted_hash"
        cls.service = "https://somesite.com"
        cls.url = BROWSE_URL
        login_url = LOGINN_URL + "/authenticate"
        login_req = dict(username=cls.user, password=cls.pwd, service=cls.service)
        r = requests.post(login_url, json=login_req)
        rsp = r.json()
        cls.token = rsp['token']
        cls.image = IMAGE

    def test_upload_browse(self):
        """ Test upload a browse """
        resource = "/browses"
        req = dict(browser=self.user, url=self.service, title="API Tests",
                   token=self.token, shot=self.image)
        r = requests.post(self.url + resource, json=req)
        self.assertEqual(r.status_code, 200)
        rsp = r.json()
        self.assertIsInstance(rsp, dict)
        self.assertIn('browser', rsp)
        self.assertIn('url', rsp)
        self.assertIn('title', rsp)
        self.assertIn('shot', rsp)
        self.assertIn('title', rsp)
        self.assertIn('published', rsp)
        self.__class__.published = rsp['published']
        self.__class__.shot = rsp['shot']

    def test_add_browser_upvote(self):
        """ Test adding upvote to browse """
        resource = "/links/upvote"
        req = dict(browser=self.user, url=self.service,
                   upvote="interesting", token=self.token)
        r = requests.post(self.url + resource, json=req)
        self.assertEqual(r.status_code, 200)
        rsp = r.json()
        self.assertIsInstance(rsp, dict)
        self.assertIn('browser', rsp)
        self.assertIn('url', rsp)
        self.assertIn('upvote', rsp)

    def test_add_browser_view(self):
        """ Test adding view to browse """
        resource = "/links/view"
        req = dict(browser=self.user, url=self.service, token=self.token)
        r = requests.post(self.url + resource, json=req)
        self.assertEqual(r.status_code, 200)
        rsp = r.json()
        self.assertIsInstance(rsp, dict)
        self.assertIn('browser', rsp)
        self.assertIn('url', rsp)

    def test_delete_browse(self):
        """ Test deleting a browse """
        r = requests.get(self.url + "/browses/" + self.user)
        self.assertEqual(r.status_code, 200)
        rsp = r.json()
        resource = "/browses"
        req = dict(browser=self.user, shot=rsp[0]['shot'],
                   token=self.token, published=rsp[0]['published'])
        r = requests.delete(self.url + resource, json=req)
        self.assertEqual(r.status_code, 200)
        rsp = r.json()
        self.assertIn('browser', rsp)
        self.assertIn('shot', rsp)
        self.assertIn('published', rsp)


class TestBrowsesUnauth(unittest.TestCase):
    """
    Test unauthenticated resources.
    """
    @classmethod
    def setUpClass(cls):
        cls.user = "joebloggs"
        cls.pwd = "salted_hash"
        cls.url = BROWSE_URL

    def test_get_latest_browses(self):
        """ Test last 24 hours of browses """
        resource = "/browses"
        r = requests.get(self.url + resource)
        rsp = r.json()
        self.assertEqual(r.status_code, 200)
        self.assertIsInstance(rsp, list)
        if len(rsp) > 0:
            self.assertIn('browser', rsp[0])
            self.assertIn('shot', rsp[0])
            self.assertIn('url', rsp[0])

    def test_get_user_browses(self):
        """ Test user browses """
        resource = "/browses/" + self.user
        r = requests.get(self.url + resource)
        rsp = r.json()
        self.assertEqual(r.status_code, 200)
        self.assertIsInstance(rsp, list)
        self.assertIn('browser', rsp[0])
        self.assertIn('shot', rsp[0])
        self.assertIn('url', rsp[0])
        self.assertIn('title', rsp[0])
        self.assertIn('browsers', rsp[0])
        self.assertIn('published_first_by', rsp[0])
        self.assertIn('published_first_time', rsp[0])
        self.assertIn('published_last_by', rsp[0])
        self.assertIn('published_last_time', rsp[0])

def main():
    unittest.main(verbosity=2)


IMAGE = """data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAA
AHgAQMAAAAPH06nAAAAA1BMVEX///+nxBvIAAAAPElEQVR42u3BAQ0AAADCIPu
nfg43YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABwKJfgA
AHF2U8TAAAAAElFTkSuQmCC"""


if __name__ == '__main__':
    main()

File: aws/apitests.js
/*
 * apitests.js
 *
 * AWS API Gateway Test Suite for Browses.
 */
const should = require('should');
const assert = require('assert');
const request = require('supertest');


describe('browses', function() {
  const url = 'https://f7mlijh134.execute-api.eu-west-1.amazonaws.com/beta';
  const s3URL = 'https://s3-eu-west-1.amazonaws.com/browses/';
  const appId = '1659456037715738';
  const secret = '7f02b4a9d73f9ca20603ace52f421158';
  const service = 'browses';
  const image = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAA
  AHgAQMAAAAPH06nAAAAA1BMVEX///+nxBvIAAAAPElEQVR42u3BAQ0AAADCIPu
  nfg43YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABwKJfgA
  AHF2U8TAAAAAElFTkSuQmCC`;
  var token = '';
  var browser = '';
  var browseID = '';
  /*
   * Get authentication token before running tests.
   */
  before(function(done) {
    request('https://graph.facebook.com/oauth')
    .get('/access_token')
    .query({ client_id: appId })
    .query({ client_secret: secret })
    .query({ grant_type: 'client_credentials' })
    .end((err, res) => {
      if (err) {
        throw err;
      }
      const accessToken = res.text.replace('access_token=', '');
      request('https://graph.facebook.com/')
      .get(`/${appId}/accounts/test-users`)
      .query({ access_token: accessToken })
      .end((err, res) => {
        if (err) {
          throw err;
        }
        const response = JSON.parse(res.text);
        browser = response.data[0].id;
        token = response.data[0].access_token;
        done();
      });
    });
  });
  /*
   * Test we can upload a browse.
   */
  describe('uploadBrowse', function() {
    describe('testUploadBrowse', function() {
      it('should return successfully with correct parameters', function(done) {
        const params = {
          token,
          url: service,
          image,
          title: 'API Tests',
        };
        request(url)
        .post('/browses')
        .send(params)
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Object);
          res.body.should.have.property('browser');
          res.body.should.have.property('name');
          res.body.should.have.property('url');
          res.body.should.have.property('title');
          res.body.should.have.property('image');
          res.body.should.have.property('published');
          browseID = res.body.image.split(s3URL)[1];
          request(s3URL)
          .get(browseID)
          .end((imgErr, imgRes) => {
            imgRes.status.should.be.equal(200);
            done();
          });
        });
      });
    });
  });
  /*
   * Test we can get the last 24 hours of browses.
   */
  describe('getLatestBrowses', function() {
    describe('testGetLatestBrowses', function() {
      it('should return successfully with correct parameters', function(done) {
        request(url)
        .get('/browses')
        .query({ page: 1 })
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Array);
          res.body[0].should.have.property('browser');
          res.body[0].should.have.property('name');
          res.body[0].should.have.property('image');
          res.body[0].should.have.property('url');
          res.body[0].should.have.property('published');
          res.body[0].should.have.property('published_first_by');
          res.body[0].should.have.property('published_first_time');
          res.body[0].should.have.property('published_last_by');
          res.body[0].should.have.property('published_last_time');
          done();
        });
      });
    });
  });
  /*
   * Test we can get a users browses.
   */
  describe('getUserBrowses', function() {
    describe('testGetUserBrowses', function() {
      it('should return successfully with correct parameters', function(done) {
        request(url)
        .get(`/browses/${browser}`)
        .query({ page: 1 })
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Array);
          res.body[0].should.have.property('browser');
          res.body[0].should.have.property('name');
          res.body[0].should.have.property('image');
          res.body[0].should.have.property('url');
          res.body[0].should.have.property('title');
          res.body[0].should.have.property('browsers');
          res.body[0].should.have.property('published_first_by');
          res.body[0].should.have.property('published_first_time');
          res.body[0].should.have.property('published_last_by');
          res.body[0].should.have.property('published_last_time');
          done();
        });
      });
    });
  });
  /*
   * Test we can add an upvote to a browse.
   */
  describe('addBrowseUpvote', function() {
    describe('testAddBrowseUpvote', function() {
      it('should return successfully with correct parameters', function(done) {
        const params = {
          token,
          url: service,
          upvote: 'interesting',
        };
        request(url)
        .post('/browses/upvote')
        .send(params)
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Object);
          res.body.should.have.property('browser');
          res.body.should.have.property('name');
          res.body.should.have.property('url');
          res.body.should.have.property('upvote');
          done();
        });
      });
    });
  });
  /*
   * Test we can add a browse view.
   */
  describe('addBrowseView', function() {
    describe('testAddBrowseView', function() {
      it('should return successfully with correct parameters', function(done) {
        const params = {
          token,
          url: service,
        };
        request(url)
        .post('/browses/view')
        .send(params)
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Object);
          res.body.should.have.property('browser');
          res.body.should.have.property('name');
          res.body.should.have.property('url');
          done();
        });
      });
    });
  });
  /*
   * Test we can delete a browse.
   */
  describe('deleteBrowse', function() {
    var browses;
    /*
     * Get browse info before running test.
     */
    before(function(done) {
      request(url)
      .get(`/browses/${browser}`)
      .end((err, res) => {
        if (err) {
          throw err;
        }
        res.status.should.be.equal(200);
        res.headers.should.have.property('access-control-allow-origin');
        res.headers['access-control-allow-origin'].should.be.equal('*');
        browses = res.body;
        done();
      });
    });

    describe('testDeleteBrowse', function() {
      it('should return successfully with correct parameters', function(done) {
        const params = {
          browser,
          published: browses[0].published,
          token,
          image: browses[0].image,
        };
        request(url)
        .delete('/browses')
        .send(params)
        .end((err, res) => {
          if (err) {
            throw err;
          }
          res.status.should.be.equal(200);
          res.headers.should.have.property('access-control-allow-origin');
          res.headers['access-control-allow-origin'].should.be.equal('*');
          res.body.should.be.instanceof(Object);
          res.body.should.have.property('browser');
          res.body.should.have.property('published');
          res.body.should.have.property('name');
          done();
        });
      });
    });
  });
});
